src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2022 Julian Kuelshammer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Julian Kuelshammer
+-/
+import data.nat.choose.central
+import algebra.big_operators.fin
+import tactic.field_simp
+import tactic.linear_combination
+
+/-!
+# Catalan numbers
+
+The Catalan numbers (http://oeis.org/A000108) are probably the most ubiquitous sequence of integers
+in mathematics. They enumerate several important objects like binary trees, Dyck paths, and
+triangulations of convex polygons.
+
+## Main definitions
+
+* `catalan n`: the `n`th Catalan number, defined recursively as
+               `catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)`.
+
+## Main results
+
+* `catalan_eq_central_binom_div `: The explicit formula for the Catalan number using the central
+  binomial coefficient, `catalan n = nat.central_binom n / (n + 1)`.
+
+## Implementation details
+
+The proof of the equality of the two definitions follows
+https://math.stackexchange.com/questions/3304415/catalan-numbers-algebraic-proof-of-the-recurrence-relation
+
+## TODO
+
+* Prove that the Catalan numbers enumerate many interesting objects.
+* Generalise to Catalan numbers associated to arbitrary Coxeter groups.
+
+-/
+
+open_locale big_operators
+open finset
+
+/-- The recursive definition of the sequence of Catalan numbers:
+`catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i)` -/
+def catalan : ℕ → ℕ
+| 0       := 1
+| (n + 1) := ∑ i : fin n.succ, have _ := i.2, have _ := nat.lt_succ_iff.mpr (n.sub_le i),
+             catalan i * catalan (n - i)
+
+@[simp] lemma catalan_zero : catalan 0 = 1 := by rw catalan
+
+lemma catalan_succ (n : ℕ) : catalan (n + 1) = ∑ i : fin n.succ, catalan i * catalan (n - i) :=
+by rw catalan
+
```

Comment:
Maybe add lemmas `catalan_one`, `catalan_two`, `catalan_three` for possible user convenience and also to convince the reader that there's no off-by-one errors?

Reply:
Seems fine to me!  Thanks for adding them.